### PR TITLE
Crud. Reviewed.

### PR DESCRIPTION
**Pull request by zasamen**
Logging in constructor. Validation of everything.

### CODE REVIEW

File: server/src/main/java/school/journal/service/CRUDService.java [hunk 1]
```diff
@@ -13,8 +13,11 @@
 public class CRUDService<T> extends ServiceAbstractClass{
     @Autowired
     protected IRepository<T> repository;
-    protected static final Logger LOGGER = Logger.getLogger(CRUDService.class);
+    protected static Logger LOGGER;
 
+    public CRUDService() {
+        LOGGER = Logger.getLogger(CRUDService.class);
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/CRUDService.java]: why does it initialize in constructor?

File: server/src/main/java/school/journal/service/IPupilService.java [hunk 2]
```diff
@@ -10,7 +9,7 @@
 
     List<Pupil> getListOfPupils(int clazzId) throws ServiceException;
 
-    Pupil movePupilToAnotherClass(Pupil pupil, Clazz clazz) throws ServiceException;
+    Pupil movePupilToAnotherClass(Pupil pupil, int classId) throws ServiceException;
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/IPupilService.java]: you can use pupilId as param in this method

File: server/src/main/java/school/journal/service/CRUDService.java (hunk 1 above)
Comment (mbarouski) [reply server/src/main/java/school/journal/service/CRUDService.java]: delete initialization

File: server/src/main/java/school/journal/service/impl/MarkService.java [hunk 3]
```diff
@@ -72,22 +79,14 @@ public Mark getOne(int id) throws ServiceException {
...
-        if (id <= 0) throw new ServiceException("Invalid Id");
-    }
-
     private void validateDate(Date date) throws ServiceException {
         Calendar c = new GregorianCalendar();
         c.add(Calendar.DAY_OF_YEAR, 1);
-        if (date.after(c.getTime())) {
-            throw new ServiceException("Invalid date");
-        }
+        if (date.after(c.getTime())) throw new ServiceException("Invalid date");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/MarkService.java]:
```
if (date.after(c.getTime())) {		        
    throw new ServiceException("Invalid date");		
}
```

File: server/src/main/java/school/journal/service/impl/MarkService.java [hunk 4]
```diff
@@ -72,22 +79,14 @@ public Mark getOne(int id) throws ServiceException {
...
     }
 
     private void validateValue(int value) throws ServiceException {
-        if (value <= 0 || value >= 11) {
-            throw new ServiceException("Invalid value");
-        }
+        if (value <= 0 || value >= 11) throw new ServiceException("Invalid value");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/MarkService.java]: the same

File: server/src/main/java/school/journal/service/impl/PupilService.java [hunk 5]
```diff
@@ -109,61 +112,45 @@ public void delete(int id) throws ServiceException {
...
-        //Plus 100year that we subtract
-        if (date.after(c.getTime())) {
-            throw new ServiceException("Invalid date");
-        }
+        if (date.before(c.getTime()))
+            throw new ServiceException("Invalid date to start education");
+        c = new GregorianCalendar();
+        c.add(Calendar.YEAR, 1);
+        if (date.after(c.getTime()))
+            throw new ServiceException("Invalid date to start education");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/PupilService.java]: use curly braces

File: server/src/main/java/school/journal/service/impl/PupilService.java [hunk 6]
```diff
@@ -109,61 +112,45 @@ public void delete(int id) throws ServiceException {
...
-            throw new ServiceException("Invalid date of starting education");
-        }
-        c.add(Calendar.YEAR,123);
-        //We have no information about pupil, who hasn't entered our school
-        //Plus 100year that we subtract
-        if (date.after(c.getTime())) {
-            throw new ServiceException("Invalid date of ending education");
-        }
+        if (date.before(c.getTime()))
+            throw new ServiceException("Invalid date to end education");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/PupilService.java]: `{}`

File: server/src/main/java/school/journal/service/impl/TeacherService.java [hunk 7]
```diff
@@ -1,28 +1,57 @@
...
-    public Teacher create(Teacher obj) throws ServiceException {
-        return null;
+
+    public Teacher create(Teacher teacher) throws ServiceException {
+        validateString(teacher.getLastName(), "Last Name");
+        validateString(teacher.getFirstName(), "First Name");
+        validateNullableString(teacher.getPathronymic(), "Patronymic");
+        validateNullableId(teacher.getClassId(), "Class");
+        validateNullableString(teacher.getDescription(), "Description");
+        MobilePhoneValidator.validate(teacher.getPhoneNumber());
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/TeacherService.java]: ok, move phone validation functionality to validation class

File: server/src/main/java/school/journal/service/impl/UserService.java [hunk 8]
```diff
@@ -9,61 +11,58 @@
...
 
-    private void checkUsername(User user) throws ServiceException {
-        if((user.getUsername() == null) || (user.getUsername().isEmpty())){
-            throw new ServiceException("Username is empty.");
-        }
+    public UserService() {
+        LOGGER = Logger.getLogger(UserService.class);
     }
 
+    private static Pattern EMAIL_PATTERN = Pattern.compile("[0-9a-zA-Z]+@[0-9a-zA-Z]");
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/UserService.java]: move email validation to `school.journal.utils.ValidateServiceUtils`

File: server/src/main/java/school/journal/service/impl/UserService.java [hunk 9]
```diff
@@ -9,61 +11,58 @@
...
         checkEmail(user);
-        user.setLocked((byte)0);
+        user.setLocked((byte) 0);
         user.setPassHash(MD5Generator.generate(generateNewPassword()));
         return super.create(user);
     }
 
     private String generateNewPassword() {
-        return MD5Generator.generate(((Long)(System.currentTimeMillis() % 1000)).toString());
+        return MD5Generator.generate(((Long) (System.currentTimeMillis() % 1000)).toString());
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/impl/UserService.java]: delete space

File: server/src/main/java/school/journal/utils/ValidateServiceUtils.java [hunk 10]
```diff
@@ -4,17 +4,21 @@
...
 
-    public static void validateString(String string,String message) throws ServiceException{
-        if( (string == null) || (string.isEmpty())){
-            throw new ServiceException(message);
-        }
+    public static void validateId(int id, String column) throws ServiceException {
+        if (id <= 0) throw new ServiceException("Invalid " + column + "Id");
+    }
+
+    public static void validateNullableString(String string, String column) throws ServiceException {
```
Comment (mbarouski) [inline server/src/main/java/school/journal/utils/ValidateServiceUtils.java]: `if (string == null) return;`

**Review by mbarouski — CHANGES_REQUESTED** [review]

File: server/src/main/java/school/journal/service/CRUDService.java (hunk 1 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/CRUDService.java]: ok. Then i should init it in the field declaration.

Comment (mbarouski) [reply server/src/main/java/school/journal/service/CRUDService.java]: no

Comment (mbarouski) [reply server/src/main/java/school/journal/service/CRUDService.java]: this field is initialized in child class

File: server/src/main/java/school/journal/service/CRUDService.java [hunk 11]
```diff
@@ -13,8 +13,11 @@
 public class CRUDService<T> extends ServiceAbstractClass{
```
Comment (mbarouski) [inline server/src/main/java/school/journal/service/CRUDService.java]: make this class abstract

File: server/src/main/java/school/journal/service/impl/MarkService.java (hunk 3 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/MarkService.java]: ok

File: server/src/main/java/school/journal/service/impl/UserService.java (hunk 8 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/UserService.java]: ok

File: server/src/main/java/school/journal/service/impl/UserService.java (hunk 9 above)
Comment (zasamen) [reply server/src/main/java/school/journal/service/impl/UserService.java]: ok

**Pull request merged**